pkg/podsvtrelabel/admission.go

```diff
@@ -0,0 +1,152 @@
+	IsExempt(request *admissionv1.AdmissionRequest) (exempt bool, response *admissionv1.AdmissionResponse)
+
+	// Admit makes an attempt to admit the specified resource in the request.
+	// It returns an AdmissionResponse that is set appropriately. On success,
+	// the response should contain the patch for update.
+	Admit(admissionSpec *admissionv1.AdmissionRequest) *admissionv1.AdmissionResponse
+}
+
+// NewInClusterAdmission returns a new instance of Admission that is appropriate
+// to be consumed in cluster.
+func NewInClusterAdmission(kubeClientConfig *restclient.Config, stopCh <-chan struct{}) (admission Admission, err error) {
+	return NewAdmission(kubeClientConfig, stopCh)
+}
+
+// NewInClusterAdmission returns a new instance of Admission that is appropriate
+// to be consumed in cluster.
+func NewAdmission(kubeClientConfig *restclient.Config, stopCh <-chan struct{}) (admission Admission, err error) {
+	client, clientErr := kubernetes.NewForConfig(kubeClientConfig)
+	if clientErr != nil {
+		err = fmt.Errorf("name=%s failed to load configuration - %s", Name, clientErr.Error())
+		return
+	}
+
+	factory := informers.NewSharedInformerFactory(client, defaultResyncPeriod)
+
+	namespaces := factory.Core().V1().Namespaces()
+	nsInformer := namespaces.Informer()
+	go nsInformer.Run(stopCh)
+
+	limitRanges := factory.Core().V1().LimitRanges()
+	limitRangeInformer := limitRanges.Informer()
+	go limitRangeInformer.Run(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, nsInformer.HasSynced) {
+		err = fmt.Errorf("name=%s failed to wait for Namespace informer cache to sync", Name)
+		return
+	}
+
+	if !cache.WaitForCacheSync(stopCh, limitRangeInformer.HasSynced) {
+		err = fmt.Errorf("name=%s failed to wait for LimitRange informer cache to sync", Name)
+		return
+	}
+
+	admission = &podSVTOverride{
+		nsLister: namespaces.Lister(),
+	}
+
+	return
+}
+
+type podSVTOverride struct {
+	nsLister corev1listers.NamespaceLister
+}
+
+func (p *podSVTOverride) IsApplicable(request *admissionv1.AdmissionRequest) bool {
+	if request.Resource.Resource == string(corev1.ResourcePods) &&
+		request.SubResource == "" &&
+		(request.Operation == admissionv1.Create || request.Operation == admissionv1.Update) {
+
+		return true
+	}
+
+	return false
+}
+
+func (p *podSVTOverride) IsExempt(request *admissionv1.AdmissionRequest) (exempt bool, response *admissionv1.AdmissionResponse) {
+	// we enforce an opt-in model.
+	// all resource(s) are by default exempt unless the containing namespace has the right label.
+	exempt = true
+
+	ns, err := p.nsLister.Get(request.Namespace)
+	if err != nil {
+		klog.Warningf("namespace=%s error retrieving namespace: %v", request.Namespace, err)
+		response = admissionresponse.WithForbidden(request, err)
+		return
+	}
+
+	enabled, exists := ns.Labels[EnabledLabelName]
+	if exists && enabled == "true" {
+		klog.V(5).Infof("namespace=%s namespace is not exempt", request.Namespace)
+
+		exempt = false
+		return
+	}
+
```

Comment:
We should check that the pod has PersistentVolumeClaim.

Reply:
I added this logic to the Admin() function... Saves one extra GetPod() call.